x/ibc/testing/coordinator.go

```diff
@@ -0,0 +1,178 @@
+// by 5 seconds.
+//
+// CONTRACT: this function must be called after every commit on any TestChain.
+func (coord *Coordinator) IncrementTime() {
+	for _, chain := range coord.Chains {
+		chain.CurrentHeader = abci.Header{
+			Height: chain.CurrentHeader.Height,
+			Time:   chain.CurrentHeader.Time.Add((timeIncrement)),
+		}
+	}
+}
+
+// CommitBlock commits a block on the provided indexes and then increments the global time.
+//
+// CONTRACT: the passed in list of indexes must not contain duplicates
+func (coord *Coordinator) CommitBlock(chains ...uint64) {
+	for _, index := range chains {
+		chain := coord.Chains[index]
+		chain.App.Commit()
+		nextBlock(chain)
+	}
+	coord.IncrementTime()
+}
+
+// CommitNBlocks commits n blocks to state and updates the block height by 1 for each commit.
+func (coord *Coordinator) CommitNBlocks(index, n uint64) {
+	chain := coord.Chains[index]
+
+	for i := uint64(0); i < n; i++ {
+		chain.App.BeginBlock(abci.RequestBeginBlock{Header: chain.CurrentHeader})
+		chain.App.Commit()
+		nextBlock(chain)
+		coord.IncrementTime()
+	}
+}
+
+// CreateClient creates a counterparty client on the source chain.
+func (coord *Coordinator) CreateClient(source, counterparty uint64) {
+	coord.CommitBlock(source, counterparty)
+	coord.Chains[source].CreateClient(coord.Chains[counterparty])
+	coord.IncrementTime()
+}
+
+// UpdateClient updates a counterparty client on the source chain.
+func (coord *Coordinator) UpdateClient(source, counterparty uint64) {
+	coord.CommitBlock(source, counterparty)
+	coord.Chains[source].UpdateClient(coord.Chains[counterparty])
+	coord.IncrementTime()
+}
+
+// CreateConnection constructs and executes connection handshake messages in order to create
+// channels on source and counterparty chains with the passed in Connection State. The
+// connectionID's of the source and counterparty are returned.
+//
+// NOTE: The counterparty testing connection will be created even if it is not created in the
+// application state.
+func (coord *Coordinator) CreateConnection(
+	sourceIndex, counterpartyIndex uint64,
+	state connectiontypes.State,
+) (sourceConnection, counterpartyConnection string) {
+	source := coord.Chains[sourceIndex]
+	counterparty := coord.Chains[counterpartyIndex]
+
+	if state == connectiontypes.UNINITIALIZED {
+		return
+	}
+
+	sourceConnection = source.NewConnection()
+	counterpartyConnection = counterparty.NewConnection()
+
+	// Initialize connection on source
+	source.ConnectionOpenInit(counterparty, sourceConnection, counterpartyConnection)
+	coord.IncrementTime()
+
+	if state == connectiontypes.INIT {
+		return
+	}
+
+	// Initialize connection on counterparty
+	counterparty.ConnectionOpenTry(source, counterpartyConnection, sourceConnection)
+	coord.IncrementTime()
+
+	if state == connectiontypes.TRYOPEN {
+		return
+	}
+
+	// Open connection on both chains
+	source.ConnectionOpenAck(counterparty, sourceConnection, counterpartyConnection)
+	coord.IncrementTime()
+
```

Comment:
Shouldn't there be this check here

```suggestion
if state == connectiontypes.ACK {
    return
}
```

Reply:
that type DNE, see [connection states](https://github.com/cosmos/cosmos-sdk/blob/08b7557fe5512cf5a0a0f8143e6b21c4d2950922/x/ibc/03-connection/types/types.pb.go#L32)